chain/client/src/client.rs

```diff
@@ -1661,4 +1669,119 @@ impl Client {
+    }
+    fn remove_old_entries(&mut self, head_height: BlockHeight) {
+        let heights_to_drop: Vec<BlockHeight> = self
+            .heights
+            .iter()
+            .take_while(|(&k, _)| k < ChunksDelayTracker::lowest_height(head_height))
+            .map(|(&k, _)| k)
+            .collect();
+        for h in heights_to_drop {
+            self.heights.remove(&h);
+        }
+    }
+
+    fn update_metrics(&mut self, head_height: BlockHeight) {
+        if head_height == self.prev_head_height {
+            return;
+        }
+        let mut cnt = 0;
+        let mut sum_delays = 0.;
+        for (&k, v) in &self.heights {
+            if k > head_height {
+                break;
+            }
+            if let Some(block_received) = v.block_received {
+                if let Some(latest_chunk) = v.chunks_received.values().max() {
+                    if latest_chunk > &block_received {
+                        cnt += 1;
+                        sum_delays +=
+                            latest_chunk.duration_since(block_received).as_micros() as f64;
+                    }
+                }
+            }
+        }
+        if cnt == 0 {
+            near_metrics::set_gauge(&metrics::CHUNKS_RECEIVING_DELAY_US, 0);
+        } else {
+            near_metrics::set_gauge(
+                &metrics::CHUNKS_RECEIVING_DELAY_US,
+                (sum_delays / (cnt as f64)).round() as i64,
+            );
+        }
+        self.prev_head_height = head_height;
+    }
+
+    fn add_block_timestamp(&mut self, height: BlockHeight, head_height: BlockHeight) {
+        self.remove_old_entries(head_height);
+        if height < ChunksDelayTracker::lowest_height(head_height) {
+            return;
+        }
+        self.heights
+            .entry(height)
+            .and_modify(|info| {
+                if info.block_received.is_none() {
+                    info.block_received = Some(Instant::now())
+                }
+            })
+            .or_insert_with(|| HeightInfo {
+                block_received: Some(Instant::now()),
+                chunks_received: Default::default(),
+            });
+        self.update_metrics(head_height);
+    }
+
+    fn add_chunk_timestamp(
+        &mut self,
+        height: BlockHeight,
+        shard_id: ShardId,
+        head_height: BlockHeight,
+    ) {
+        self.remove_old_entries(head_height);
+        if height < ChunksDelayTracker::lowest_height(head_height) {
+            return;
+        }
+        self.heights
+            .entry(height)
+            .and_modify(|info| {
+                info.chunks_received.entry(shard_id).or_insert_with(|| Instant::now());
+            })
+            .or_insert_with(|| {
+                let mut chunks_received = HashMap::new();
+                chunks_received.insert(shard_id, Instant::now());
+                HeightInfo { block_received: None, chunks_received }
+            });
```

Comment:
Seems like you can replace and_modify/or_insert_with pair with a single `.or_deafult().chuns_received.entry(shrad_id).or_insert_with(Instant::now)`